### PR TITLE
Adding block_ordinal to BlockHeaderInnerRest

### PR DESCRIPTION
**Pull request by Kouprin**
This change is necessary for enabling Epoch Sync https://github.com/near/nearcore/issues/3488.

It creates BlockHeaderV3 and BlockHeaderInnerRestV3 and a nightly `protocol_feature_block_ordinal` feature. The actual Block Ordinal of the new Block is (the amount of Blocks already existing on the Canonical Chain + 1) which is taken from Bowen-Merkle Trees.

Tested with `test_block_ordinal` sanity test and CI.

### CODE REVIEW

**Comment by Kouprin** [comment]
`test_cache_invalidation` passes locally. Is github cargo-test configured specifically for current version with not including upgrade?

File: core/primitives/src/version.rs [hunk 1]
```diff
@@ -52,6 +52,8 @@ pub const DELETE_KEY_STORAGE_USAGE_PROTOCOL_VERSION: ProtocolVersion = 40;
 
 pub const SHARD_CHUNK_HEADER_UPGRADE_VERSION: ProtocolVersion = 41;
 
+pub const BLOCK_ORDINAL_UPGRADE_VERSION: ProtocolVersion = 45;
```
Comment (frol) [inline core/primitives/src/version.rs:55]: @bowenwang1996 what is the current process for the protocol features? Should all of them go through nightly? This version 45 while the stable is 41, and the latest nightly is 47 looks off to me.

**Review by frol — APPROVED** [review]
view-client and primitives look good to me :+1:

File: chain/chain/src/chain.rs [hunk 2]
```diff
@@ -3199,10 +3199,6 @@ impl<'a> ChainUpdate<'a> {
             return Err(ErrorKind::InvalidChunkMask.into());
         }
 
-        if !header.verify_chunks_included() {
-            return Err(ErrorKind::InvalidChunkMask.into());
```
Comment (bowenwang1996) [inline chain/chain/src/chain.rs:3203]: why is this removed?

**Review by bowenwang1996 — CHANGES_REQUESTED** [review]
New features should go through nightly first. See https://github.com/near/NEPs/issues/126#issuecomment-712130812 for the idea and you can take a look at https://github.com/near/nearcore/blob/191559126134075b3d5476410ab061b0f6b8509d/core/primitives/src/version.rs#L78 to see how it is done

File: chain/chain/src/chain.rs (hunk 2 above)
Comment (Kouprin) [reply chain/chain/src/chain.rs]: I wrongly assumed this check is not necessary because it's always true. Reverted.

File: core/primitives/src/version.rs (hunk 1 above)
Comment (Kouprin) [reply core/primitives/src/version.rs]: Added the feature to nightly.

File: core/primitives/src/version.rs [hunk 3]
```diff
@@ -173,3 +177,42 @@ macro_rules! checked_feature {
         }
     }};
 }
+
+#[macro_export]
+macro_rules! checked_feature_crate {
+    ($feature_name:tt, $feature:ident, $current_protocol_version:expr) => {{
+        #[cfg(feature = $feature_name)]
+        let is_feature_enabled = crate::version::PROTOCOL_FEATURES_TO_VERSION_MAPPING
+            [&crate::version::ProtocolFeature::$feature]
```
Comment (frol) [inline core/primitives/src/version.rs:186]: Is the only change here `near_primitives::` being renamed into `crate::`? What is the issue you have faced with the original macro?

Comment (Kouprin) [reply core/primitives/src/version.rs:186]: It has no `near_primitives::` and unable to compile.

Comment (frol) [reply core/primitives/src/version.rs:186]: I see. We need to use `$crate` instead of `near_primitives`: https://doc.rust-lang.org/1.5.0/book/macros.html#the-variable-crate

Comment (Kouprin) [reply core/primitives/src/version.rs]: It seems $crate works perfectly. Thank you for the insight. Updated.

**Review by frol — APPROVED** [review]

File: core/primitives/src/block_header.rs [hunk 4]
```diff
@@ -434,6 +543,36 @@ impl BlockHeader {
...
+        } else {
+            let inner_rest = BlockHeaderInnerRestV3 {
+                chunk_receipts_root,
+                chunk_headers_root,
+                chunk_tx_root,
+                challenges_root,
+                random_value: CryptoHash::default(),
+                validator_proposals: vec![],
+                chunk_mask: vec![true; chunks_included as usize],
+                block_ordinal: 1, // TODO MOO Bowen please make sure genesis is block #1 in terms of Merkle Proofs
```
Comment (bowenwang1996) [inline core/primitives/src/block_header.rs:555]: this is the case

File: core/primitives/src/block_header.rs [hunk 5]
```diff
@@ -563,6 +719,16 @@ impl BlockHeader {
...
             BlockHeader::BlockHeaderV1(header) => &header.inner_rest.chunk_mask,
             BlockHeader::BlockHeaderV2(header) => &header.inner_rest.chunk_mask,
+            BlockHeader::BlockHeaderV3(header) => &header.inner_rest.chunk_mask,
+        }
+    }
+
+    #[inline]
+    pub fn block_ordinal(&self) -> NumBlocks {
+        match self {
+            BlockHeader::BlockHeaderV1(_) => 0, // not applicable
```
Comment (bowenwang1996) [inline core/primitives/src/block_header.rs:729]: Should we return an option then? It is better to take advantage of the rust type system than to rely on some hardcoded constant.

File: core/primitives/src/views.rs [hunk 6]
```diff
@@ -415,6 +416,7 @@ pub struct BlockHeaderView {
     pub random_value: CryptoHash,
     pub validator_proposals: Vec<ValidatorStakeView>,
     pub chunk_mask: Vec<bool>,
+    pub block_ordinal: NumBlocks,
```
Comment (bowenwang1996) [inline core/primitives/src/views.rs:419]: this should ideally also guarded by the feature flag. While not doing so has no impact on the protocol level, rpc users may get confused as to what this is.

File: core/primitives/src/block_header.rs (hunk 4 above)
Comment (Kouprin) [reply core/primitives/src/block_header.rs]: I've updated the comment.

File: core/primitives/src/block_header.rs (hunk 5 above)
Comment (Kouprin) [reply core/primitives/src/block_header.rs:729]: We can do that.
However, BlockHeader is very basic primitive that is used everywhere in the code. Once we use Option<NumBlocks> for small piece of semantical sugar here, we will have to unwrap it everywhere for forever. I wouldn't go this way as it may create tons of `match ...` and `unwrap()` code later with no actual purpose.

File: core/primitives/src/views.rs (hunk 6 above)
Comment (Kouprin) [reply core/primitives/src/views.rs:420]: Updated.

File: core/primitives/src/block_header.rs (hunk 5 above)
Comment (Kouprin) [reply core/primitives/src/block_header.rs:729]: We can introduce Option wrapper for block_ordinal in BlockHeaderView only. It makes sense to me. Should we go this way?

Comment (bowenwang1996) [reply core/primitives/src/block_header.rs:729]: Yeah let's do that

**Review by bowenwang1996 — APPROVED** [review]

File: pytest/lib/messages/block.py [hunk 7]
```diff
@@ -266,6 +286,28 @@ class ApprovalInner:
             ]
         }
     ],
+    [
+        BlockHeaderInnerRestV2, {
```
Comment (Kouprin) [inline pytest/lib/messages/block.py:290]: This is a typo.
